File: VodArchiver/VodArchiver/VideoInfo/HitboxVideoInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace VodArchiver.VideoInfo {
	[Serializable]
	public class HitboxVideoInfo : IVideoInfo {
		HitboxVideo VideoInfo;

		public HitboxVideoInfo( HitboxVideo video ) {
			VideoInfo = video;
		}

		public override XmlNode Serialize( XmlDocument document, XmlNode node ) {
			node.AppendAttribute( document, "_type", "HitboxVideoInfo" );
			node.AppendChild( VideoInfo.Serialize( document, document.CreateElement( "VideoInfo" ) ) );
			return node;
		}

		public override StreamService Service {
			get {
				return StreamService.Hitbox;
			}

			set {
				throw new InvalidOperationException();
			}
		}

		public override string Username {
			get {
				return VideoInfo.MediaUserName;
			}

			set {
				VideoInfo.MediaUserName = value;
			}
		}

		public override string VideoGame {
			get {
				return VideoInfo.MediaGame;
			}

			set {
				VideoInfo.MediaGame = value;
			}
		}

		public override string VideoId {
			get {
				return VideoInfo.MediaId.ToString();
			}

			set {
				VideoInfo.MediaId = int.Parse( value );
			}
		}

		public override TimeSpan VideoLength {
			get {
				return TimeSpan.FromSeconds( VideoInfo.MediaDuration );
			}

			set {
				VideoInfo.MediaDuration = value.TotalSeconds;
			}
		}

		public override RecordingState VideoRecordingState {
			get {
				return RecordingState.Recorded;
			}

			set {
				throw new InvalidOperationException();
			}
		}

		public override DateTime VideoTimestamp {
			get {
				return VideoInfo.MediaDateAdded;
			}

			set {
				VideoInfo.MediaDateAdded = value;
			}
		}

		public override string VideoTitle {
			get {
				return VideoInfo.MediaTitle;
			}

			set {
				VideoInfo.MediaTitle = value;
			}
		}

		public override VideoFileType VideoType {
			get {
				if ( VideoInfo.MediaProfiles.First().Url.EndsWith( "m3u8" ) ) {
					return VideoFileType.M3U;
				}

				return VideoFileType.Unknown;
			}

			set {
				throw new InvalidOperationException();
			}
		}
	}
}

File: native/vodarchiver/videoinfo/hitbox-video-info.cpp
#include "hitbox-video-info.h"

#include <cstdint>
#include <format>
#include <stdexcept>
#include <string>
#include <string_view>
#include <utility>

#include "util/number.h"

namespace VodArchiver {
HitboxVideoInfo::~HitboxVideoInfo() = default;

StreamService HitboxVideoInfo::GetService() const {
    return StreamService::Hitbox;
}

std::string_view HitboxVideoInfo::GetUsername(std::array<char, 256>& buffer) const {
    return this->VideoInfo.MediaUserName;
}

std::string_view HitboxVideoInfo::GetVideoGame(std::array<char, 256>& buffer) const {
    return this->VideoInfo.MediaGame;
}

std::string_view HitboxVideoInfo::GetVideoId(std::array<char, 256>& buffer) const {
    auto result = std::format_to_n(buffer.data(), buffer.size() - 1, "{}", this->VideoInfo.MediaId);
    *result.out = '\0';
    return std::string_view(buffer.data(), result.out);
}

TimeSpan HitboxVideoInfo::GetVideoLength() const {
    return TimeSpan::FromDoubleSeconds(this->VideoInfo.MediaDuration);
}

RecordingState HitboxVideoInfo::GetVideoRecordingState() const {
    return RecordingState::Recorded;
}

DateTime HitboxVideoInfo::GetVideoTimestamp() const {
    return this->VideoInfo.MediaDateAdded;
}

std::string_view HitboxVideoInfo::GetVideoTitle(std::array<char, 256>& buffer) const {
    return this->VideoInfo.MediaTitle;
}

VideoFileType HitboxVideoInfo::GetVideoType() const {
    if (!VideoInfo.MediaProfiles.empty() && VideoInfo.MediaProfiles.front().Url.ends_with("m3u8")) {
        return VideoFileType::M3U;
    }
    return VideoFileType::Unknown;
}

std::unique_ptr<IVideoInfo> HitboxVideoInfo::Clone() const {
    return std::make_unique<HitboxVideoInfo>(*this);
}
} // namespace VodArchiver
